### PR TITLE
Fix classification cache_ram leak via shared tensor

### LINKED ISSUE #9824: High RAM  using when  add (cache=True)in  Classification Training
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

_No response_

### Bug

when i add  (cache=True)in  Classification Training , the Ram using is increasing every epoch , until it crash the training , start like from 3   to 6  to 11 to 15 ....... 50 , GB
but if i don't add it , the ram using work fine , it be like 4 GB and all training is fixed 

i work on colab 
`!yolo task=classify mode=train cache=True model=yolov8n-cls.pt data='/content/Classification-1' epochs=5  batch=265 imgsz=128 `
  

### Environment

_No response_

### Minimal Reproducible Example

_No response_

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by raimbekovm**
## Summary

Fixes #9824 — `ClassificationDataset` `cache='ram'` leaks RAM linearly when `num_workers > 0` because the per-sample lazy write-back triggers per-worker copy-on-write on `fork` and full-pickle bloat on `spawn`. The legacy workaround force-disabled `cache_ram` with a warning.

This PR replaces lazy write-back with a single shared `torch.uint8` buffer that is preloaded once on the main process, pinned via `share_memory_()` before workers are forked, and read-only from the worker perspective:

- `check_cache_ram` — sample N images, estimate raw RAM with safety margin (mirrors `BaseDataset.check_cache_ram` style); guards against empty datasets
- `cache_images` — two-pass design:
  - **Pass 1** `Image.open(f).size` (header read only, no decode) records `(h, w, 3)` per sample → compute total bytes and per-image offsets
  - **Pass 2** allocate `torch.empty(total, dtype=uint8)` once, then `cv2.imread` + slice-copy each image into its offset via `ThreadPool(NUM_THREADS)`
  - Pin with `share_memory_()` so all workers map the same POSIX SHM region
  - Wrapped in `try/except (MemoryError, OSError, RuntimeError)` — gracefully falls back to disk reads with a warning naming `--shm-size` if `/dev/shm` quota (e.g. Docker default 64 MB) blocks `share_memory_()`
- `__getitem__` reads `cache[offset[i] : offset[i] + h*w*c].reshape(h, w, c)` — original-size view, so augmentations (`RandomResizedCrop` train, `Resize+CenterCrop` val) run downstream exactly as with `cache=False`

`safety_margin=1.0` (≈ 2× cache size budget) accommodates streaming/heap-fragmentation overhead during preload, measured at ~2.2× peak factor on macOS.

## Validation

| Layer | Setup | Result |
|---|---|---|
| Smoke | `yolo classify train` on imagenet10 | trains correctly, ~2× faster than `cache=False` |
| Semantic parity | val + training transforms vs `cache=False` | 0/6 tensor mismatches (originally 6/6 with pre-resized cache) |
| 6-epoch stress | 15K imgs, 224, batch=64, workers=4 | drift +22 MB (`cache=ram`) vs +21 MB (`cache=False`) — leak gone |
| Edge cases (×4) | SHA-256 of cache before/after iter | byte-identical: re-iter, non-square resize, train/val separation, heavy augment (randaugment+erasing+hsv ×5 epochs) |
| DDP simulation | `mp.spawn(N=2)` | within-rank shared (1× cache); cross-rank duplication (N×) is expected DDP semantics, not a regression |
| Empty dataset | `samples = []` | `check_cache_ram` returns False, no crash |
| `share_memory_()` failure | mocked OSError | fallback works, `__getitem__` returns valid tensors |
| 30-epoch long-run | 15K imgs, ~2.5 h, 120+ worker spawn cycles | drift −42 MB e1→e30, USS oscillates in 97 MB band on 1.3 GB baseline |

## Test plan

- [x] `yolo classify train data=imagenet10 cache=ram epochs=2` completes
- [x] `cache=ram` does not grow with `num_workers`
- [x] `cache=ram` byte-identical to `cache=False` across iteration / augmentation / train+val
- [x] DDP-style multi-process spawn works (each rank preloads independently)
- [x] Empty dataset does not crash
- [x] `share_memory_()` failure falls back gracefully (mocked)
- [x] No regression in `cache=False` and `cache=disk` paths (untouched code paths)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
This PR overhauls export testing and documentation 📦🧪 by adding isolated CI environments for model exports, expanding Python 3.13 support, and refreshing deployment/export docs across ONNX, RKNN, DEEPX, Hailo, and Ultralytics Platform integrations.

### 📊 Key Changes
- Added a new CI helper script, `.github/scripts/create-export-env.py`, to automatically build and smoke-test isolated export environments from the exporter registry 🔧
- Introduced a new **`IsolatedExports`** GitHub Actions job so heavy or specialized export formats run in their own virtual environments instead of bloating the main CI pipeline 🚀
- Updated core CI to:
  - move default testing to **Python 3.13**
  - use `export-base` instead of the larger `export` dependency set
  - disable auto-install during tests for more predictable runs
  - improve cache handling and asset preloading for slow tests ⚙️
- Split export dependencies in `pyproject.toml` into more targeted groups like:
  - `export-base`
  - `export-tensorflow`
  - `export-coreml`
  - `export-executorch`
  - `export-deepx`
  - `export-legacy-torch` 📚
- Expanded and refined export test coverage:
  - added **ONNX INT8 export** testing
  - added **RKNN export** testing
  - improved isolation for export artifacts to avoid parallel test conflicts
  - added checks to ensure each export format is mapped to a valid environment ✅
- Improved Docker cleanup by pruning build cache and cleaning package caches to reduce image bloat and CI disk usage 🐳
- Added major new documentation for **Hailo HEF workflows**, explaining how to export YOLO models through ONNX and compile externally with Hailo tooling 📖
- Expanded docs for **ONNX INT8**, **RKNN INT8**, **DEEPX**, **ExecuTorch**, **QNN**, **TensorRT**, and other integrations with clearer support notes and updated examples 🧠
- Added new **Ultralytics Platform integrations** docs covering imports from **Ultralytics HUB** and **Roboflow** 🔗
- Improved docs build tooling so generated pages now get meaningful default descriptions and keywords instead of placeholder text ✍️
- Refreshed various examples, naming, links, and wording across the docs for accuracy and consistency, including better DEEPX branding and export guidance 🛠️

### 🎯 Purpose & Impact
- Makes export testing more reliable by isolating fragile or conflicting dependencies, reducing CI breakage across formats like DEEPX, ExecuTorch, RKNN, and others 🧪
- Speeds up and simplifies the main test pipeline by keeping specialized export stacks out of the shared environment ⏱️
- Improves compatibility with newer Python versions, especially **Python 3.13**, helping users and contributors stay current 🐍
- Gives developers clearer dependency boundaries, making export-related installs easier to manage and debug 📦
- Expands confidence in advanced deployment workflows through stronger automated validation of export formats ✅
- Provides much better guidance for users deploying models to edge and hardware-specific targets like **Hailo**, **Rockchip RKNN**, **DEEPX**, and **ONNX INT8** devices 🌍
- Improves discoverability and usability of **Ultralytics Platform** migration/import features for teams bringing data in from HUB or Roboflow 📥
- Overall, this PR strengthens both the **engineering quality** and **user-facing documentation** around exporting and deploying Ultralytics models 💡

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you shortly. Please review the checklist below:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good direction overall: the shared tensor preload should address the per-worker copy problem cleanly. I found two issues that need attention before merging: an empty-dataset crash in the RAM-fit check, and a functional regression where `cache_ram` now changes classification preprocessing by squashing images to `imgsz` during preload.

💬 Posted 2 inline comments

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The shared-tensor approach addresses the original worker copy-on-write leak cleanly, but there are two resource-management regressions to fix before merging: preload currently doubles peak RAM usage, and unconditional `share_memory_()` can fail in environments with limited shared-memory quota even when regular RAM is sufficient.

💬 Posted 2 inline comments

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The shared-memory cache approach looks clean and well-scoped. I did not find any clear correctness, performance, or maintainability issues in the diff based on the visible context.

File: ultralytics/data/dataset.py [hunk 1]
```diff
@@ -788,6 +794,95 @@ def __len__(self) -> int:
         """Return the total number of samples in the dataset."""
         return len(self.samples)
 
+    def check_cache_ram(self, safety_margin: float = 0.5) -> bool:
```
Comment (Y-T-G) [inline ultralytics/data/dataset.py:1050]: You can replace this and the one in `BaseDataset` with a common function and move it to data/utils.py since they are the same 

https://github.com/ultralytics/ultralytics/blob/0ba1d385f8bc589f4049d27a96b02afc4ab87feb/ultralytics/data/base.py#L324

**Comment by Y-T-G** [comment]
I think this spawn bug also exists in the `BaseDataset` caching method. So we should later fix that in a new PR. We could probably unify the RAM caching method then to be the same. You can add a TODO.

**Comment by raimbekovm** [comment]
Superseded by #24667, recreated on a clean branch so the diff is reviewable (this branch accumulated many merge-from-main commits while open). Closing in favor of #24667.
